### ISSUE TITLE
[UX] Adding a person to a team

### ISSUE THREAD

**Issue by karai17** [ludumdare/ludumdare#1157]
I think the UX for adding someone to a team is a bit non-obvious. My suggestion would be to have the team UI in the game page itself instead of your user profile. When creating or editing a game, the Team section would have a list of team members with you at the top plus an Add Team Member button that creates a dropdown populated with your mutual friends.

## Team

* Karai
* holo ( [remove]() )
* `_______________| v |`
* Add member +

**Comment by povrazor** [comment]
Yep, the current system sucks. I didn't have a working drop-down box until this past Saturday.

**Comment by karai17** [comment]
Erm, can you explain that a little more? I do web programming as well as game programming and I am very curious about these sorts of things. I'm sure you're about sick of my criticism by now, but it all comes from a place of interest and love, as I value LD very much and want to see things get better, not worse.

**Comment by povrazor** [comment]
The UX is built from scratch on top of Preact. The super fancy modern way of doing web stuff these days is with a Virtual DOM. Libraries like React, Preact, Inferno, and many others have this concept of a duplicate DOM that is compared against the real DOM. The VDOM is extremely light, without all the fat of the real DOM. And like the idea of "Dirty Rectangles" in graphics programming, only the things that change need to be updated.

So even basic controls like buttons or text boxes need to be reimplemented. You make a class out of them, and that class can be instanced like `<MyDropDownBox arg="true"></MyDropDownBox>`, in a JS document with special features. This is also no longer about raw JavaScript, but rather compiled JavaScript.

Here's an example:

https://github.com/ludumdare/ludumdare/blob/master/src/com/button-base/base.js

In this case we've corrected a bug you typically see when using an SVG element inside a div as a button. That's one of the nice part is we can actually fix bugs in HTML itself, and make the components we use more logical. No need to remember that checkboxes use the `value` attribute, textboxes the child, or whatever. You can fix HTML itself by wrapping it in a VDOM.

**Comment by karai17** [comment]
This seems like an extremely over-engineered solution compared to "put an html dropdown in a template and populate it with some server-side script". Every language has its quirks, "fixing" them isn't always the best use of time imo. HTML sucks in various ways, but I am definitely not a fan of the direction JS development has gone. It feels very much like "Enterprise Java", if you are aware of that meme.

**Comment by povrazor** [comment]
Maintaining large software projects is hard. The bigger a project is, the more there is to fix, to break, and the harder it is to get in to. Once you pass a certain complexity, you need practices and tools to help you keep things manageable. The problem is, if you don't take time to clean up after, to document, then it gets nightmarish. The mistake of enterprise is educating and hiring people that don't know how to do that, or not valuing those that can enough. It's not even a meme. Java as a language was ruined by success.

**Comment by karai17** [comment]
What you said is very true, but that doesn't really address my comment. Yes, big, complex applications need good documentation, unit tests, a strict code style, and all that good stuff. What you didn't address is that you as a programmer (or a team) should be knowledgable about the technologies you are using to build your thing. Is HTML weird? Yes. Is HTML *broken*? No. It has quirks and legacy cruft like any other aged technology and people working with it should be aware of those quirks. Writing an abstraction layer or a shim on top of every oddity you find in every language you use would make your codebase itself the most significant barrier to entry for anyone wanting to contribute. Instead of being a web developer, anyone wanting to help out needs to become a Ludum Dare developer before they can even attempt to write a single line of code.

**Comment by povrazor** [comment]
> Instead of being a web developer, anyone wanting to help out needs to become a Ludum Dare developer before they can even attempt to write a single line of code.

Well we're doing better now than we ever have. The source code has been open source since Phil and I took over a decade ago, but not until there was a custom toolchain and documentation did we get outside contributions.

**Comment by karai17** [comment]
The only thing worse than trying to get into an over-engineered codebase is getting into a Wordpress/php codebase ;P

Anyway, I'm not trying to critisize the code too much, I haven't taken much time to look into it to have a strong opinion one way or the other. I'm am just of the opinion that if something can be done simpler, it ought to be. Learning to work within the quirks of the tools I use is something I've learned the hard way. Every time I try to go against the grain, it causes me a lot more time and stress than just throwing in the towel and accepting that the world isn't a perfect sphere. I find your design choice of not having a single line of HTML within the code base of a web application to be... odd.

The only stab I'll take at you is that this website has been in the works for quite a while and is only now just barely functional enough to actually use. Since last LD, the only appreciable difference I can note is that you fixed the caching issue that had made the site *actually* unusable. I don't know if you've been writing a metric tonne of behind-the-scenes boiler plate that, when finished, will make all the rest of the features of the site fall into place like magic, but from a user's perspective, not much seems to be going on, and perhaps looking into simpler solutions to some problems would have made the site come together a lot faster.

**Comment by Jezzamonn** [comment]
At the risk of getting further sidetracked, something worth noting is that
the LD site isn't unique in terms of how it handles the front-end stuff.
It's basically done using react, Facebook's front end framework. (Except
instead it's a smaller version called preact). And backend php is pretty
normal.

It'd be interesting in hearing the design decisions involved in picking
those frameworks, but yeah probably a discussion for a different place

**Comment by povrazor** [comment]
For what it's worth, I started developing it in a fashion like you suggested (i.e. no dependencies). Ended up throwing it away and starting over as it was getting too fat and too difficult to maintain (though you can find remnants of the old codebase under `/sandbox`). I took a few weeks and studied the options available, brushed up on the standard, and the setup we have now is what made the most sense to me.

It may sound like we're reinventing the wheel, but we're not. Dirty Rectangles and Double Buffering are some of the oldest concepts in programming. We're building a nearly native site on top of a thin layer (the VDOM). Almost every major online service is doing exactly what we're doing (we're literally using a derivative of Facebook). The code is still JavaScript, but it's ES2017 Javascript (i.e. JS according to the latest industry standard spec). 

This is the modern standard agreed upon by Google, Mozilla and Microsoft. Every feature we use like Classes, Promises, Async operations, this is all what Javascript is now. The only non-standard thing we're using is JSX, but that's fast become and industry standard, even if it's not part of the spec. 

For JSX support and compatibility reasons the code is run through a compiler. And hey, Internet Explorer still happens. At least for the next little while we need to continue to support it.  If we compiled away the JSX, the website could be run natively in Firefox or Chrome. But if we really cared, the site could be made to run on Internet Explorer versions from over a decade ago. What we're doing is not weird. :wink:

> Since last LD, the only appreciable difference I can note is

Sorry. I don't know if you've ever crunched on a project you love for months at a time, on very little sleep, stressed out of your mind with a handful of people insulting you and your efforts, who are disguised by idea of you making a living from something you've sunk over a decade of your life in to, gave up hobbies, shut down your business, and gone in to debt over. I needed some time to recover. :wink:

People don't come out of the woodwork to help you make stuff until you have a foundation. To reach that foundation milestone, I literally had to recreate WordPress and bunch of extensions from the ground up. Knowing this had to be an improvement on top of WordPress, there was a lot of careful planning that went in to it too. And when we finally had that foundation, some amazing people like @zwrawr, @Noojuno, @StormBurpee showed up and started helping, and now @sgstair and @Jezzamonn too. It was hard road to get here, but I'm so glad we made it. Yes it's not perfect, but the reaction has been so much more positive this time around. 

I'm proud that the machine of improvement is working. It's been years since I had help on the technical side (since Phil left). I could get help with data entry and analysis, but nobody would step up fix a bug. I'm so much happier with how things are going today. I've needed help on the tech side for so long. :smile:

**Comment by karai17** [comment]
I'm glad things are working out for sure. There are definitely shitty people out there who criticize success, be it from envy or otherwise. None of my criticism comes from that. I have no problem with you or anyone else earning a living, or even becoming wealthy, by making something cool that people enjoy. I think a lot of criticism from myself, and others in the community, come from what on the surface can look like hubris or ego. LD38 was the first use of this new platform and it was *extremely* painful to the point where I personally gave up using the website and expected my game to not qualify. We ended up qualifying with something like 12 or 14 reviews in each category which, considering your earlier comment in another issue about how you tweak the minimum on a per-LD basis, suggests a lot of people gave up and the minimum had to be set quite low to meet some respectable threshold of qualifying games.

While the new platform is improving, there is still a lot of animosity in the community. Lots of features are still not functioning or not functioning properly or up to par, the website has a lot of oddities such as buttons that don't do anything or messages that tell you some feature isn't there yet, and the preceived quality just feels lower. The old, bad website worked for the users, and the new, shiny website kind of works, sometimes, for some users. At the end of the day, people are upset or angry because **we love Ludum Dare** and it can feel at times that ego has overtaken community as a priority. Whether or not that is true is less important than the perception that it is true. There definitely feels like a disconnect between the administration and the community these days, and that worries some people.

As for the technical side, I don't think "no dependencies" is something I would claim is good or useful. There is no point to re-inventing the wheel for the sake of a pure, self-contained code base. I just don't really like the direction web programming is heading. For my projects, I use Lapis (written by @leafo originally for itch.io, but turned into its own thing) and it's a wonderful framework that has all the bells and whistles a typical web application needs. It clearly scales well as I've mentioned earlier today that itch runs on a $40 Linode and is extremely responsive. Both approaches work.

tl;dr Keep doing you, but don't forget to think about what your users actually want/need, as opposed to what might be preceived as a want or a need.

**Comment by povrazor** [comment]
> It'd be interesting in hearing the design decisions involved in picking those frameworks, but yeah probably a discussion for a different place

Well hehe, we've already gone very off topic (I'll just have to fix this so we can close the bug after, heh).

I come from a gamedev background. I've shipped more than a dozen games on variety of systems, notably in this case: the GameBoy. Pretty early in my career I became a bit of a nut about performance and file size, because we had to be. We had kilobytes to work with, not megabytes. I've shipped games written in Assembly, because we had to.

Fast forward through 20-some years of indie/homebrew dev, many years in the industry, back again as an indie but one with professional experience. I've warmed up to modern dev ideas, but I always have a watchful eye out for where I can make things smaller or squeeze out more performance.

PHP is a piece of software I know how to get great performance out of. It's what we used on the old LD website (WordPress), so part of the reason is because it's what I know. The PHP team has made some HUGE strides in general performance leading up to the release of PHP 7, so while PHP 7 was in beta I decided to start using it. And though I haven't needed to for the LD site, I've done experiments with native C code and PHP. So I know how I can offload work to C code if really need it. We shouldn't though. I'm not sure our servers have GPUs anyway.

The important point however is shared memory. The standard way to share data with OTHER web languages is via a Memcached server. It's a separate application you connect to via TCP or a Socket, and write key/value pairs to. PHP though, it has a library with an awkward history named APCu. Generally speaking, it's a key/value store like Memcached, but it's written to use shared memory directly. No need to pack your requests and responses in to TCP packets or whatever you want to call Socket data. You read/write memory directly, using proper shared memory locking. I've literally edited the code myself, so I know how it's working.

So with PHP and APCu, I can outperform any Web Language+Memcached setup. That's why it's the backend (and Node JS).

Node JS would have been the other option, but though Node is excellent at doing many things on one core, it by-design scales poorly across multiple cores. That said it's still an excellent tool. Most of our toolchain comes from Node JS. 

Long story short, "I know how to get **the most** performance from PHP". 

Continuing on, modern JavaScript was chosen instead of Ruby or others because I like native languages. Assembly is the native language of the CPU. With enough knowledge, you can write C code and know exactly what assembly it's generating (if you see me writing some weird looking `for` loops, I'm sorry, that's me optimal C programmer coming out). On the web, the lowest level you can go is JavaScript, so the code is JavaScript.

Babel has a 100k'ish runtime. Babel gives you nearly 100% ES2017 support, but me Mr file size fanatic doesn't want to pay 100k for that. Especially given that Buble will do many of those things for free. No addon library required. ES2017 language features that are missing get converted to features that available (i.e. arrow functions to normal functions). I had to make a few small tweaks to Preact for Buble compatibility, and the author accepted them to mainline. So now mainline Preact supports Buble.

@Noojuno and I have started adding features to the compiler. We're still working on getting Async/Await compatibility entirely there, but it's functional today, and the overhead is nothing. Nodent, the addon library we're using for legacy async/await support does add a bit of code, but it builds on top of Javascript Promises. It's very much in the same ballpark as Buble (only what's needed as its needed).

So we're using Buble because I like what it stands for, and I've warmed up to working on the compiler itself directly, so if we really want something we can have it.

Typical web stuff think 100k is "lightweight". As someone that came from the GameBoy, that's an insult to me. That said there are "factions" of web developers that agree with me, that lightweight means 1 kb (or less). The guy behind Preact, we see eye to eye. I swear it's also not because he lives like hour away from me. :wink: Preact is a very powerful abstraction library, and it's nearly 100% compatible with React, one of the most popular frameworks today (also skipped because of a 100k runtime). It's fast too. There technically are 2 faster frameworks in the same vein, but the creator is awesome. He helped me get started with it, and has been great about my patches for Buble support among other things. And it just keeps getting better. The list of sites that have deployed Preact frontends in the past 2-3 months mind boggling. So it's light weight, it's extremely well supported, and it's being proven again and again. I liked it when I found it, I had no idea so many others did too.

I think that's everything weird. ;)

Vagrant and Virtual Box because I can make a virtual server run on both Linux and Windows. Can't do that with Docker.

**Comment by povrazor** [comment]
> tl;dr Keep doing you, but don't forget to think about what your users actually want/need, as opposed to what might be preceived as a want or a need.

Yep. That's why we need to keep running Ludum Dare's. :wink:

I get a lot of support e-mails and messages. When I get the same message again and again, that's how I know we have a serious problem. Of course lots of things would be nice, but if people are hitting roadblocks, those are things that need to be resolved sooner than later. 

I can't predict everything that's going to be a problem for people. I've literally canned features because of things I've learned doing support (they would have been discarded anyway). You can't develop something as large as Ludum Dare in a vacuum. Like a hurricane it just has to happen. We'll pick up the pieces as quickly and as best we can.
